### PR TITLE
BUG: fix concat of Sparse with non-sparse dtypes

### LINKED ISSUE #34336: REGR: concat of Sparse with incompatible dtype now gives Sparse[object] instead of object
Regression on master related to the `common_dtype` mechanism, I suppose:

```
In [5]: pd.__version__    
Out[5]: '1.0.3'

In [6]: s1 = pd.Series([1, 0, 2], dtype=pd.SparseDtype("int64", 0)) 

In [7]: s2 = pd.Series(["a", "b", "c"], dtype="category") 

In [8]: pd.concat([s1, s2]) 
Out[8]: 
0    1
1    0
2    2
0    a
1    b
2    c
dtype: object

In [9]: pd.concat([s2, s1]) 
Out[9]: 
0    a
1    b
2    c
0    1
1    0
2    2
dtype: object
```

(and the same on v0.25.3)

But on master:

```
# raising in SparseDtype._get_common_dtype
In [3]: pd.concat([s1, s2])  
...
TypeError: data type not understood

In [4]: pd.concat([s2, s1])   
Out[4]: 
0    a
1    b
2    c
0    1
1    0
2    2
dtype: Sparse[object, 0]
```

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #34336

This adds tests with the behaviour as it was on 0.25.3 / 1.0.3, and some changes to get back to that behaviour.  
(but whether this behaviour is fully "sane", I am not sure ..)

### CODE REVIEW

File: pandas/core/dtypes/concat.py [hunk 1]
```diff
@@ -81,6 +81,12 @@ def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
         except ValueError:
             return arr.astype(object, copy=False)
 
+    if is_sparse(arr) and not is_sparse(dtype):
```
Comment (TomAugspurger) [inline pandas/core/dtypes/concat.py:84]:
```suggestion
    if is_sparse(arr.dtype) and not is_sparse(dtype):
```

**Review by TomAugspurger — COMMENTED** [review]
Looks good.

I'm not totally sure about the expected behavior of `concat([Sparse, Categorical])`. I suspect it was never properly discussed / intentionally implemented. Happy to just match 1.0.3 behavior for now htough.

**Comment by jorisvandenbossche** [comment]
OK, will go forward with this PR as is (matching the previous behaviour) to unblock the other concat PRs, but will open a set of follow-up issues.

File: pandas/core/arrays/sparse/array.py [hunk 2]
```diff
@@ -1063,7 +1063,8 @@ def astype(self, dtype=None, copy=True):
         """
         dtype = self.dtype.update_dtype(dtype)
         subtype = dtype._subtype_with_str
-        sp_values = astype_nansafe(self.sp_values, subtype, copy=copy)
+        # TODO copy=False is broken for astype_nansafe with int -> float
+        sp_values = astype_nansafe(self.sp_values, subtype, copy=True)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/sparse/array.py:1067]: Opened https://github.com/pandas-dev/pandas/issues/34456 for this (and added link in the comment)

File: pandas/core/dtypes/concat.py [hunk 3]
```diff
@@ -81,6 +81,12 @@ def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
         except ValueError:
             return arr.astype(object, copy=False)
 
+    if is_sparse(arr.dtype) and not is_sparse(dtype):
+        # problem case: SparseArray.astype(dtype) doesn't follow the specified
+        # dtype exactly, but converts this to Sparse[dtype] -> first manually
+        # convert to dense array
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/concat.py:87]: Opened https://github.com/pandas-dev/pandas/issues/34457 for this

**Comment by jorisvandenbossche** [comment]
And opened https://github.com/pandas-dev/pandas/issues/34459 for the general "what should concat(sparse, categorical) do?" question

File: pandas/core/dtypes/concat.py [hunk 4]
```diff
@@ -20,7 +21,7 @@
 )
 from pandas.core.dtypes.generic import ABCCategoricalIndex, ABCRangeIndex, ABCSeries
 
-from pandas.core.arrays import ExtensionArray
+from pandas.core.arrays import ExtensionArray, SparseArray
```
Comment (TomAugspurger) [inline pandas/core/dtypes/concat.py:24]: Might be creating a circular import? https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=36392&view=logs&j=bef1c175-2c1b-51ae-044a-2437c76fc339&t=770e7bb1-09f5-5ebf-b63b-578d2906aac9

Comment (jorisvandenbossche) [reply pandas/core/dtypes/concat.py]: Yeah, I already pushed a fix. 
I don't really understand why, though, as `SparseArray` is included in the `pandas.core.arrays` init, just as ExtensionArray.

**Pull request merged**